cmake/keccak.cmake

```diff
@@ -29,20 +29,268 @@ if (NOT __KECCAK_INCLUDED)
+    "if(POLICY CMP0076)\n"
+    "  cmake_policy(SET CMP0076 NEW)\n"
+    "endif()\n"
+    "\n"
+    "# Kerl Optimization\n"
+    "option(KERL_REF_32 \"Kerl reference 32bit\" OFF)\n"
+    "option(KERL_AsmX86_64 \"Kerl optimized AsmX86-64\" OFF)\n"
+    "option(KERL_COMPACT \"Kerl minimizing code and memory sizes for 64bit\" OFF)\n"
+    "option(KERL_AVX2 \"Kerl with AVX2 instruction set optimization\" OFF)\n"
+    "option(KERL_INPLACE_32 \"Kerl minimizing memory sizes for 32bit\" OFF)\n"
+    "option(KERL_64u6 \"Kerl for 64-bit platforms with 6 rounds unrolled\" OFF)\n"
+    "option(KERL_64uf \"Kerl for 64-bit platforms with full rounds unrolled\" OFF)\n"
+    "option(KERL_T2o1 \"Kerl calls Keccak-p[1600] twice\" OFF)\n"
+    "option(KERL_T4o1 \"Kerl calls Keccak-p[1600] four times\" OFF)\n"
+    "option(KERL_T4o2 \"Kerl calls Keccak-p[1600]x2 twice\" OFF)\n"
+    "option(KERL_T8o1 \"Kerl calls Keccak-p[1600] eight times\" OFF)\n"
+    "option(KERL_T8o2 \"Kerl calls Keccak-p[1600]x2 four times\" OFF)\n"
+    "option(KERL_T8o4 \"Kerl calls Keccak-p[1600]x4 twice\" OFF)\n"
+    "# Kerl times 2 SIMD\n"
+    "option(KERL_SIMD128_SSE2u2 \"Kerl SSE2 instruction set, with two rounds unrolled\" OFF)\n"
+    "option(KERL_SIMD128_SSE2uf \"Kerl SSE2 instruction set, with full rounds unrolled\" OFF)\n"
+    "# Kerl times 4 SIMD\n"
+    "option(KERL_SIMD256_AVX2u6 \"Kerl AVX2 instruction set, with 6 rounds unrolled\" OFF)\n"
+    "option(KERL_SIMD256_AVX2u12 \"Kerl AVX2 instruction set, with 12 rounds unrolled\" OFF)\n"
+    "option(KERL_SIMD256_AVX2uf \"Kerl AVX2 instruction set, with full rounds unrolled\" OFF)\n"
+    "# Kerl ARM Optimization\n"
+    "option(KERL_INPLACE_ARMv6M \"Kerl minimizing memory sizes for ARMv6M\" OFF)\n"
+    "option(KERL_INPLACE_ARMv7M \"Kerl minimizing memory sizes for ARMv7M\" OFF)\n"
+    "option(KERL_INPLACE_ARMv7A \"Kerl minimizing memory sizes for ARMv7A\" OFF)\n"
+    "option(KERL_ARMv7A_ASM \"Kerl neon optimiaztion for ARMv7A\" OFF)\n"
+    "option(KERL_ARMv8A_ASM \"Kerl neon optimiaztion for ARMv8A\" OFF)\n"
+    "\n"
+    "if(KERL_T8o4)\n"
+    "  if((NOT KERL_T4o2) AND (NOT KERL_T4o1))\n"
+    "    message(FATAL_ERROR \"KERL_T8o4 needs KERL_T4o2 or KERL_T4o1\")\n"
+    "  endif()\n"
+    "endif()\n"
+    "\n"
+    "if(KERL_T8o2)\n"
+    "  SET(KERL_T2o1 ON CACHE BOOL FORCE \"Kerl calls Keccak-p[1600] twice\")\n"
+    "endif()\n"
+    "\n"
+    "if(KERL_T4o2)\n"
+    "  SET(KERL_T2o1 ON CACHE BOOL FORCE \"Kerl calls Keccak-p[1600] twice\")\n"
+    "endif()\n"
+    "\n"
+    "set(my_src ${keccak_src_dir}/lib/high/Keccak/KeccakSpongeWidth1600.c\n"
     "${keccak_src_dir}/lib/high/Keccak/FIPS202/KeccakHash.c)\n"
     "add_library(keccak STATIC \${my_src})\n"
+    "\n"
+    "if(KERL_AsmX86_64)\n"
+    "  target_compile_options(keccak PUBLIC -fomit-frame-pointer -O2 -g0 -march=native -mtune=native)\n"
+    "  target_sources(keccak PUBLIC ${keccak_src_dir}/lib/low/KeccakP-1600/OptimizedAsmX86-64/KeccakP-1600-x86-64-gas.s)\n"
```

Comment:
Is it possible to automatically generate this part? Many of these files were moved and/or renamed recently (https://github.com/XKCP/XKCP/commit/4bbc04d3393bc25e1eab7da7f47730446b9ea1a0)

Reply:
unfortunately, XKCP uses a different build generator that means it will cost too much time to integrate with CMake and cannot guarantee it will work on the next release. a better way would be https://github.com/XKCP/XKCP/issues/39